samples/shrubbery/parse.cc

```diff
@@ -0,0 +1,332 @@
+#include "shrubbery.h"
+#include "wf.h"
+
+// Shrubbery notation constructs that are not supported:
+// - Single quotes as opener/closer pairs
+// - Line and column insensitivity with << and >>
+// - Block comments with #// //#
+// - @-notation
+// - Keywords prefixed by ~
+//
+// Other things that could be implemented:
+// - Continuing a line with backslash
+// - Better parsing of strings
+// - Numbers other than integers
+
+namespace shrubbery
+{
+  // An Indent is a source code location where an indentation level has been
+  // established. A line at some indentation level can also be continued if the
+  // next line is more indented and starts with an operator. For example, the
+  // code
+  //
+  // f(1) + 2
+  //   + 3
+  //   - 4
+  //
+  // is identical to `f(1) + 2 + 3 - 4`. The plus on the second line can
+  // arbitrarily indented, but the minus must be in the same column as the plus
+  // on the previous line. When parsing the program above, the Indent will be
+  // ((0,0), 2), indicating that the established indentation level is line 0
+  // colum 0, and that this line is continued on column 2.
+  struct Indent {
```

Comment:
As a general note on Trieste usage, there are two strategies that may make your life considerably easier as regards meaningful whitespace. The first, and perhaps simplest, is to make the parser even dumber and have it explicitly capture whitespace as `Whitespace` and `NewLine` tokens, i.e.:

```c++
  R"(\r?\n)" >> [](auto& m){
    m.add(NewLine);
  }

  R"([ \t]+)" >> [](auto& m){
    m.add(Whitespace);
  }
```

You can then defer decisions about indents etc. to a dedicated pass in Trieste, where you'll have much more flexibility and context to make your decisions. This works rather well. Another option is to use much the same pattern, but to use pattern matching to measure the indent explicitly while processing the end of the previous line, like so: 

```c++
  R"(\r?\n([ \t]*))" >> [](auto& m){
    size_t indent = m.match(1).len;
    // your indent logic
  }
```

This lookahead can save a bunch of bookkeeping (you can preemptively push or pop indent blocks).

Reply:
I was thinking about a version with the explicit whitespace tokens, but tried writing the Shrubbery parser as a way to stress the parser module more than the existing samples (when this was developed, infix was the only other sample in the repo). It is also nice how, modulo whitespace sensitivity, the Trieste parser sort of looks like a generalized shrubbery parser. The second option does look a lot nicer than what I am doing right now though, so I will give that a spin!